### PR TITLE
Add support for enum.Enum based configuration options

### PR DESCRIPTION
**Pull request by romainletendart**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -147,6 +147,15 @@ Features
...
         config.getpath('django.static_root', pathlib.Path(BASE_DIR / 'static'))
 
+    ``getconf`` can also convert options to user-defined standard-type-based types:
+
+    .. code-block:: python
+
+        class Environment(str, enum.Enum):
+            DEV = 'dev'
+            PROD = 'prod'
+        config.getenum('environment', Environment.PRODUCTION)
```
Comment (xavfernandez) [inline README.rst:157]:
```suggestion
        config.getenum('environment', Environment.PROD)
```
I guess ?

File: getconf/base.py [hunk 2]
```diff
@@ -192,6 +193,31 @@ def getpath(self, key, default=Path('.'), doc=''):
             logger.exception("Unable to cast %r as Path for the key %s.", value, key)
             raise
 
+    def getenum(self, key, default=None, doc='', enum_class=None):
+        if enum_class is None and isinstance(default, enum.Enum):
+            enum_class = type(default)
+        assert (
+            default is not None and enum_class is not None and issubclass(enum_class, enum.Enum) and issubclass(enum_class, type(default))
+        ), 'getenum("%s", %r, enum_class=%r) is missing a default value or this one has an invalid type.' % (
```
Comment (xavfernandez) [inline getconf/base.py:201]: I don't think the default value should be mandatory if an `enum_class` is provided.

Comment (xavfernandez) [reply getconf/base.py:201]: Maybe we should stick with `None` in such case then ?
```
In [9]: CONFIG.getstr('foo')
Out[9]: ''

In [10]: CONFIG.getstr('foo', default=None)

In [11]: CONFIG.getint('foo')
Out[11]: 0

In [12]: CONFIG.getint('foo', default=None)
```

To follow the same behavior, I'd expect `CONFIG.getenum('foo', default=None, enum_class=TestIntEnum)` to return `None` if no configured value is found.

I agree that we would be missing a consistent behavior for `CONFIG.getenum('foo', enum_class=TestIntEnum)` but also returning `None` in such case seems like an acceptable default.

**Review by RemiCardona — APPROVED** [review]
LGTM :+1:

File: docs/reference.rst [hunk 3]
```diff
@@ -80,6 +80,14 @@ The ``BaseConfigGetter`` class
 
         Retrieve a key from available configuration sources, and parse it as a pathlib.Path object.
 
+    .. py:method:: getenum(key, default[, enum_class])
```
Comment (xavfernandez) [inline docs/reference.rst:83]:
```suggestion
    .. py:method:: getenum(key, [default, enum_class])
```

File: docs/reference.rst [hunk 4]
```diff
@@ -80,6 +80,14 @@ The ``BaseConfigGetter`` class
...
+
+        Retrieve a key from available configuration sources, and parse it as an enum.Enum based object.
+
+        .. note:: The ``default`` param accepts either an enum.Enum based instance or an enum.Enum member value type.
```
Comment (xavfernandez) [inline docs/reference.rst:87]:
```suggestion
        .. note:: The ``default`` param accepts either an enum.Enum based instance, an enum.Enum member value type or None.
```

**Review by xavfernandez — APPROVED** [review]

**Comment by RemiCardona** [comment]
@romainletendart Still LGTM

**Pull request merged**